### PR TITLE
Chapter 6

### PR DESCRIPTION
**Pull request by devtuananh**

### CODE REVIEW

File: app/assets/stylesheets/application.scss [hunk 1]
```diff
@@ -12,7 +12,4 @@
  *
  *= require_tree .
  *= require_self
- */
-
-@import "bootstrap-sprockets";
-@import "bootstrap";
\ No newline at end of file
+ */
```
Comment (mdmahmudurrahman) [inline app/assets/stylesheets/application.scss]: Please add a line break at the end of the file.

Comment (devtuananh) [reply app/assets/stylesheets/application.scss]: Yes, I've edited

File: app/helpers/application_helper.rb [hunk 2]
```diff
@@ -1,12 +1,12 @@
...
+        base_title                                      
       else
-        page_title + " | " + base_title                 # String concatenation
+        page_title + " | " + base_title                 
       end
     end
-  end
\ No newline at end of file
+  end
+
```
Comment (mdmahmudurrahman) [inline app/helpers/application_helper.rb]: Add a line break at the end of the file

File: app/models/user.rb [hunk 3]
```diff
@@ -0,0 +1,10 @@
+class User < ApplicationRecord
+  before_save { self.email = email.downcase }
+  validates :name, presence: true, length: { maximum: 50 }
+  VALID_EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-.]+\.[a-z]+\z/i
+  validates :email, presence: true, length: { maximum: 255 },
+  format: { with: VALID_EMAIL_REGEX },
+  uniqueness: { case_sensitive: false }
+  has_secure_password
+  validates :password, presence: true, length: { minimum: 6 }
+end
```
Comment (mdmahmudurrahman) [inline app/models/user.rb]: Add a line break at the end of the file

File: app/views/layouts/_footer.html.erb [hunk 4]
```diff
@@ -1,13 +1,13 @@
 <footer class="footer">
   <small>
-    The <a href="http://www.railstutorial.org/">Ruby on Rails Tutorial</a>
-    by <a href="http://www.michaelhartl.com/">Michael Hartl</a>
+  The <a href="http://www.railstutorial.org/">Ruby on Rails Tutorial</a>
```
Comment (mdmahmudurrahman) [inline app/views/layouts/_footer.html.erb]: 1. When you place something inside the parent tag, please use 2 space indentation
2. Please study about **I18n(Internationalization)** in rails and implement it in your application
Reference: [http://guides.rubyonrails.org/i18n.html](http://guides.rubyonrails.org/i18n.html)

File: app/views/layouts/_footer.html.erb [hunk 5]
```diff
@@ -1,13 +1,13 @@
...
+  by <a href="http://www.michaelhartl.com/">Michael Hartl</a>
   </small>
   <nav>
-    <ul>
-      <li><%= link_to "About",   about_path %></li>
-      <li><%= link_to "Contact", contact_path %></li>
-      <li><a href="http://news.railstutorial.org/">News</a></li>
-    </ul>
+  <ul>
+  <li><%= link_to "About", about_path %></li>
```
Comment (mdmahmudurrahman) [inline app/views/layouts/_footer.html.erb]: 1. Indentation
2. I18n 
Reference: [http://guides.rubyonrails.org/i18n.html](http://guides.rubyonrails.org/i18n.html)

File: app/views/layouts/_header.html.erb [hunk 6]
```diff
@@ -1,12 +1,12 @@
...
-    <nav>
-      <ul class="nav navbar-nav navbar-right">
-        <li><%= link_to "Home",    root_path %></li>
-        <li><%= link_to "Help",    help_path %></li>
-        <li><%= link_to "Log in", '#' %></li>
-      </ul>
-    </nav>
+   <%= link_to "sample app", root_path, id: "logo" %>
+  <nav>
+  <ul class="nav navbar-nav navbar-right">
```
Comment (mdmahmudurrahman) [inline app/views/layouts/_header.html.erb]: indentation

File: app/views/layouts/_header.html.erb [hunk 7]
```diff
@@ -1,12 +1,12 @@
...
+  <li><%= link_to "Home", root_path %></li>
```
Comment (mdmahmudurrahman) [inline app/views/layouts/_header.html.erb]: same here

File: app/views/layouts/application.html.erb [hunk 8]
```diff
@@ -1,18 +1,18 @@
...
   </head>
   <body>
-    <%= render 'layouts/header' %>
-    <div class="container">
-      <%= yield %>
-      <%= render 'layouts/footer' %>
-    </div>
+  <%= render "layouts/header" %>
+  <div class="container">
+  <%= yield %>
```
Comment (mdmahmudurrahman) [inline app/views/layouts/application.html.erb]: indentation

File: config/routes.rb [hunk 9]
```diff
@@ -1,8 +1,8 @@
...
-  get  '/contact', to: 'static_pages#contact'
-  get  '/signup',  to: 'users#new'
-end
\ No newline at end of file
+  get "users/new"
+  root "static_pages#home"
+  get  "/help", to: "static_pages#help"
+  get  "/about", to: "static_pages#about"
+  get  "/contact", to: "static_pages#contact"
+  get  "/signup", to: "users#new"
```
Comment (mdmahmudurrahman) [inline config/routes.rb]: Only 1 space between the parameters

File: test/models/user_test.rb [hunk 10]
```diff
@@ -0,0 +1,7 @@
+require 'test_helper'
```
Comment (mdmahmudurrahman) [inline test/models/user_test.rb]: Please remove this auto generated test file.

File: test/fixtures/users.yml [hunk 11]
```diff
@@ -0,0 +1,9 @@
+# Read about fixtures at http://api.rubyonrails.org/classes/ActiveRecord/FixtureSet.html
```
Comment (mdmahmudurrahman) [inline test/fixtures/users.yml]: Please remove this auto generated test file.

File: app/assets/stylesheets/custom.scss [hunk 12]
```diff
@@ -1,108 +1,100 @@
...
 /* typography */
 
-h1,
-h2,
-h3,
-h4,
-h5,
-h6 {
-    line-height: 1;
+h1,h2,h3,h4,h5,h6 {
```
Comment (mdmahmudurrahman) [inline app/assets/stylesheets/custom.scss]: 1 space between the tags

File: app/helpers/application_helper.rb [hunk 13]
```diff
@@ -1,12 +1,12 @@
 module ApplicationHelper
-
-    # Returns the full title on a per-page basis.       # Documentation comment
-    def full_title(page_title = '')                     # Method def, optional arg
-      base_title = "Ruby on Rails Tutorial Sample App"  # Variable assignment
-      if page_title.empty?                              # Boolean test
-        base_title                                      # Implicit return
+     
+    def full_title(page_title = '')                     
+      base_title = "Ruby on Rails Tutorial Sample App"
```
Comment (mdmahmudurrahman) [inline app/helpers/application_helper.rb]: Please study about **I18n(Internationalization)** in rails and implement it in your application
Reference: [http://guides.rubyonrails.org/i18n.html](http://guides.rubyonrails.org/i18n.html)

File: app/views/layouts/_footer.html.erb [hunk 14]
```diff
@@ -1,13 +1,13 @@
 <footer class="footer">
   <small>
-    The <a href="http://www.railstutorial.org/">Ruby on Rails Tutorial</a>
-    by <a href="http://www.michaelhartl.com/">Michael Hartl</a>
+  The <a href="http://www.railstutorial.org/">Ruby on Rails Tutorial</a>
+  by <a href="http://www.michaelhartl.com/">Michael Hartl</a>
```
Comment (mdmahmudurrahman) [inline app/views/layouts/_footer.html.erb]: I18n 
Reference: [http://guides.rubyonrails.org/i18n.html](http://guides.rubyonrails.org/i18n.html)

File: app/views/layouts/_footer.html.erb [hunk 15]
```diff
@@ -1,13 +1,13 @@
...
   </small>
   <nav>
-    <ul>
-      <li><%= link_to "About",   about_path %></li>
-      <li><%= link_to "Contact", contact_path %></li>
-      <li><a href="http://news.railstutorial.org/">News</a></li>
-    </ul>
+  <ul>
+  <li><%= link_to "About", about_path %></li>
+  <li><%= link_to "Contact", contact_path %></li>
```
Comment (mdmahmudurrahman) [inline app/views/layouts/_footer.html.erb]: I18n

File: app/views/layouts/_header.html.erb [hunk 16]
```diff
@@ -1,12 +1,12 @@
...
+  <li><%= link_to "Help", help_path %></li>
+  <li><%= link_to "Log in", "#" %></li>
```
Comment (mdmahmudurrahman) [inline app/views/layouts/_header.html.erb]: Implement I18n when there a plain text.

File: app/views/layouts/application.html.erb [hunk 17]
```diff
@@ -1,18 +1,18 @@
 <!DOCTYPE html>
 <html>
   <head>
-    <title><%= full_title(yield(:title)) %></title>
-    <%= csrf_meta_tags %>
-    <%= stylesheet_link_tag    'application', media: 'all',
-                                              'data-turbolinks-track': 'reload' %>
-    <%= javascript_include_tag 'application', 'data-turbolinks-track': 'reload' %>
-    <%= render 'layouts/shim' %>
+  <title><%= full_title(yield(:title)) %></title>
```
Comment (mdmahmudurrahman) [inline app/views/layouts/application.html.erb]: Remove the optional parentheses
`
full_title yield(:title)
`

File: app/views/static_pages/about.html.erb [hunk 18]
```diff
@@ -8,4 +8,4 @@
   to teach web development with
   <a href="http://rubyonrails.org/">Ruby on Rails</a>.
   This is the sample application for the tutorial.
```
Comment (mdmahmudurrahman) [inline app/views/static_pages/about.html.erb]: I18n

File: app/views/static_pages/help.html.erb [hunk 19]
```diff
@@ -6,4 +6,4 @@
   To get help on this sample app, see the
   <a href="http://www.railstutorial.org/book"><em>Ruby on Rails Tutorial</em>
   book</a>.
```
Comment (mdmahmudurrahman) [inline app/views/static_pages/help.html.erb]: I18n

File: app/views/static_pages/home.html.erb [hunk 20]
```diff
@@ -1,14 +1,11 @@
...
-
-  <%= link_to "Sign up now!", '#', class: "btn btn-lg btn-primary" %>
+  <%= link_to "Sign up now!", "#", class: "btn btn-lg btn-primary" %>
 </div>
-
-<%= link_to image_tag("rails.png", alt: "Rails logo"),
-            'http://rubyonrails.org/' %>
\ No newline at end of file
+  <%= link_to image_tag("rails.png", alt: "Rails logo"),
+  "http://rubyonrails.org/" %>
```
Comment (mdmahmudurrahman) [inline app/views/static_pages/home.html.erb]: 1. If you make intentional line break, please use 2 space indentation to the next line
2. I18n

File: app/assets/stylesheets/custom.scss [hunk 21]
```diff
@@ -1,108 +1,100 @@
...
-    margin-bottom: 30px;
-    text-align: center;
-    font-weight: normal;
-    color: #777;
+  font-size: 1.2em;
+  letter-spacing: -1px;
+  margin-bottom: 30px;
+  text-align: center;
+  font-weight: normal;
+  color: #777;
```
Comment (mdmahmudurrahman) [inline app/assets/stylesheets/custom.scss]: Use color variable and import it from an external file.

File: app/assets/stylesheets/custom.scss [hunk 22]
```diff
@@ -1,108 +1,100 @@
...
+  text-transform: uppercase;
+  letter-spacing: -1px;
+  padding-top: 9px;
+  font-weight: bold;
 }
 
 #logo:hover {
-    color: #fff;
-    text-decoration: none;
+  color: #fff;
```
Comment (mdmahmudurrahman) [inline app/assets/stylesheets/custom.scss:69]: Use color variable and import it from an external file.

File: app/assets/stylesheets/custom.scss [hunk 23]
```diff
@@ -1,108 +1,100 @@
...
 
 footer {
-    margin-top: 45px;
-    padding-top: 5px;
-    border-top: 1px solid #eaeaea;
-    color: #777;
+  margin-top: 45px;
+  padding-top: 5px;
+  border-top: 1px solid #eaeaea;
+  color: #777;
```
Comment (mdmahmudurrahman) [inline app/assets/stylesheets/custom.scss]: Same here

File: app/assets/stylesheets/custom.scss [hunk 24]
```diff
@@ -1,108 +1,100 @@
...
 }
 
 footer a {
-    color: #555;
+  color: #555;
 }
 
 footer a:hover {
-    color: #222;
+  color: #222;
```
Comment (mdmahmudurrahman) [inline app/assets/stylesheets/custom.scss]: Same here

**Pull request merged**